engine/src/main/kotlin/io/provenance/scope/classloader/MemoryClassLoader.kt

```diff
@@ -45,19 +59,13 @@ class MemoryClassLoader(
             return classCache[name]!!
         }
 
+        val rootClassName = name.split('$').first()
```

Comment:
The list of classes in ProtoHash doesn't include all of the child classes of protos that get generated. This split ensures we are just looking at the top-level java class name... Not doing this led to issues when running a contract where the child classes were in a different class loader than the parent (app vs memory), and would lead to functions getting skipped due to class comparison mismatch.

Reply:
Hmm. One thing that would be good to verify is if this works for both the java and kotlin generated concrete classes of ContractHash/ProtoHash. An example would be the `example-java` vs. `example-kotlin` sub projects in the p8e-gradle-plugin repo. I only say that because I know from looking at the files in the past I thought I recall them having different full path class representations. It would just be good to verify this work for both of them.

Reply:
Good idea, I'll try this out

Reply:
I did some manual verification, and the java contracts seem to work fine. It seems that the java contracts are written nested in a class, so they do end up with the $ divider. I added a specific test to reflect this, as well as updated the kotlin test ProtoHash to reflect the $ being where it would be in a bootstrap-generated class.

Reply:
Great!